url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/quantum/SystemMatrix.java

```diff
@@ -0,0 +1,233 @@
+		
+		for (int i = 0; i < mNumbQubits; i++)
+				clone.measured[i] = measured[i];
+	
+		return clone;
+	}
+
+	public void setCoeff(final MathsItem c, int... indeces) throws Exception {
+		if (indeces.length != mNumbQubits)
+			throw new Exception("Not enough indeces supplied");
+
+		int id = 0;
+		for (int i = 0; i < mNumbQubits; i++)
+			if (indeces[i] != 0)
+				id += indeces[i] * (1 << i);
+		coeff[id] = c;
+
+	}
+
+	public MathsItem getCoeff(int... indeces) throws Exception {
+		if (indeces.length != mNumbQubits)
+			throw new Exception("Not enough indeces supplied");
+
+		int id = 0;
+		for (int i = 0; i < mNumbQubits; i++)
+			if (indeces[i] != 0)
+				id += indeces[i] * (1 << i);
+		return coeff[id];
+	}
+
+	public int getIdFromIndexes(int... indeces) throws Exception {
+		if (indeces.length != mNumbQubits)
+			throw new Exception("Not enough indeces supplied");
+
+		int id = 0;
+		for (int i = 0; i < mNumbQubits; i++)
+			if (indeces[i] != 0)
+				id += indeces[i] * (1 << i);
+		return id;
+	}
+
+	public static int[] getIndexesFromId(int id, int mNumbQubits) {
+
+		int[] idx = new int[mNumbQubits];
+		for (int i = 0; i < mNumbQubits; i++)
+			idx[mNumbQubits - i - 1] = id - ((id = (id >> 1)) << 1);
+
+		return idx;
+	}
+
+	@Override
+	public String toString() {
+		String rep = "";
+
+		for (int i = 0; i < size; i++) {
+			
+			if (coeff[i].isZero())
+				continue;
+			
+			if (i != 0) rep += " + \n";
+			
+			coeff[i].simplify();
+			rep += coeff[i]+"\t"+getBraKet(i);
+		}
+
+		return rep;
+	}
+	
+	public String getBraKet(int id) {
+		final int[] idxes = getIndexesFromId(id, mNumbQubits);
+		String rep = "|";
+
+			for (int j = 0; j < idxes.length; j++)
+				if (!measured[j])
+					rep += idxes[j];
+				else
+					rep += "_";
+
+			rep += ">";
+		return rep;
+	}
+	
+	public static String getBraKet(int id, int mNumbQubits) {
+		final int[] idxes = getIndexesFromId(id, mNumbQubits);
+		String rep = "|";
+
+			for (int j = 0; j < idxes.length; j++)
+					rep += idxes[j];
+
+			rep += ">";
+		return rep;
+	}
+	
+	public String printValues(final HashMap<String, Complex> rules, final boolean normalize) {
+		String rep = "";
+		final double norm = normalize ? Math.sqrt(getQuickProbability(rules)) : 1.0d;
+		boolean first = true;
+
+		for (int i = 0; i < size; i++) {
+			
+			if (coeff[i].isZero())
+				continue;
+			
+			final Complex numb = Complex.divide(coeff[i].getValue(rules),norm);
+			if (numb.isZero())
+				continue;
+			
+			if (!first)
+				rep += " + \n";
+			else
+				first = false;
+			
+			rep += numb+"\t"+getBraKet(i);
+		}
+
+		return rep;
+	}
+	
+	public void performReverse(Operator ... actions) throws Exception
+	{
+		for (int i = actions.length - 1; i >= 0; i--) {
+			if (!Tools.SILENT) Tools.logger.println("Performing "+actions[i]);
+			actions[i].operate(this);
+			if (!Tools.SILENT && Tools.VERBOSE) {
+				(new AddCoeffTogether()).operate(this);
+				Tools.logger.println(this);
+				Tools.logger.println();
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 186:194, 199:207.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.